### PR TITLE
feat: can add route handler as callable

### PR DESCRIPTION
**Pull request by kenjis**
**Description**
- You can jump to the controller method with IDE.

E.g.:
```php
$routes->get('home', [Hello::class, 'index']);
$routes->get('product/(:num)/(:num)', [[Hello::class, 'index'], '$1/$2']);
```

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by mostafakhudair** [comment]
Please include additions in the user guide

**Comment by iRedds** [comment]
``[[Home::class, 'index'], '$1/$2']`` doesn't look nice

What about?
```php
//[Home::class, 'index', '$1/$2'] 

[$class, $method, $params] = $to; 
is_callable([$class, $method], true, $callableName)

//... 
```php

**Comment by kenjis** [comment]
`[Home::class, 'index', '$1/$2']` seems to be nice.
But we cannot jump to `Home::index()` source code when we click `'index'`

Screenshot 2022-02-21 9 18 29
Screenshot 2022-02-21 9 18 44

**Comment by iRedds** [comment]
Yes, you are right.

**Review by paulbalandan — COMMENTED** [review]
This needs writeup in the docs.

**Comment by iRedds** [comment]
In theory. Can we drop ``$1/$2``?

**Comment by kenjis** [comment]
I will write the docs later.

**Comment by kenjis** [comment]
Added User Guide.

**Comment by iRedds** [comment]
@kenjis can you reply to the comment above?
I want to clarify. I meant that ``$1/$2`` is useless in my opinion.
```php
$routes->get('product/(:num)/(:num)', [Hello::class, 'index']);
```
Here we already have two arguments ``(:num)/(:num)``. Which can be passed to a controller method without the ``$1/$2`` notation.
The same principle as the closure works.
```php
$routes->get('product/(:num)/(:num)', fn ($param1, $param2) => 'do something');
```
i.e. we don't specify ``$1/$2`` and parameters are passed as is.

**Comment by lonnieezell** [comment]
This looks like a great addition.  I would be in favor of removing the $1/$2 params as being specifically needed. While I've had times in the past where I needed to reorder or leave out some parameters, it's never been a big enough deal that it couldn't be worked around, and would clean this up a bit. Maybe make it optional? If it's not provided it defaults to the order given?

**Comment by iRedds** [comment]
To skip the parameter, the developer can use non-capturing grouping.
For example, for the url ``product/1/2`` and the route ``product/(?::num)/(:num)``, the parameter will only be ``2``.

**Comment by kenjis** [comment]
@iRedds @lonnieezell how about this?

**Comment by iRedds** [comment]
It will not work correctly with a non-capturing group.
For ``product/(?::num)/(:num)`` the replacement string will be ``controller::method/$1/$2`` and we get the string ``controller::method/1/`` which gives parameter [1, ''].
Perhaps with nested groups there will be a similar situation.

It seems to me that if you use your approach, then you need to trim the slash in ``Router::checkRoutes()``.
Or you can add parameters directly to ``Router::checkRoutes()``
```php
} elseif (strpos($val, '$') === false && count($matches) > 1) {
    array_shift($matches);
    $val.= '/' . implode('/', $matches);
}

$this->setRequest(explode('/', $val));
```
This solution will also allow parameters to be applied to more than just the $to array. 
For example, ``$routes->get('x/(:num)', 'Home::index');``
But this is a quick example and I could miss the nuances.

**Comment by kenjis** [comment]
I think regex is originally complex, and we do not need to support all the pattern.

- do not support non-capturing groups
- do not support nested groups

Or I don't mind to change the implementation if there is simple better one.

But if `$routes->get('x/(:num)', 'Home::index');` will add params automatically,
it is a breaking change.

**Comment by kenjis** [comment]
Updated the documentation: https://github.com/codeigniter4/CodeIgniter4/pull/5713/files#diff-2a34c841c28af79b5550af42e3e5c24b337fa9ce9e90f8164000ac6c7d5ea7fe

**Review by lonnieezell — APPROVED** [review]
Looks great. Thanks!

File: system/Router/RouteCollection.php [hunk 1]
```diff
@@ -1227,6 +1231,52 @@ protected function create(string $verb, string $from, $to, ?array $options = nul
         }
     }
 
+    /**
+     * @return string
+     */
+    private function processArrayCallableSyntax(string $from, array $to)
```
Comment (paulbalandan) [inline system/Router/RouteCollection.php:1237]:
```suggestion
    private function processArrayCallableSyntax(string $from, array $to): string
```

File: system/Router/RouteCollection.php [hunk 2]
```diff
@@ -1227,6 +1231,52 @@ protected function create(string $verb, string $from, $to, ?array $options = nul
...
+        preg_match_all('/\(.+?\)/', $from, $matches);
+        $count = is_countable($matches[0]) ? count($matches[0]) : 0;
+
+        $params = '';
+
+        if ($count > 0) {
+            for ($i = 1; $i <= $count; $i++) {
+                $params .= '/$' . $i;
+            }
+        }
```
Comment (paulbalandan) [inline system/Router/RouteCollection.php:1275]: No need for this if. If count is 0, the for will not execute.

**Review by paulbalandan — APPROVED** [review]

**Comment by kenjis** [comment]
@paulbalandan The commit you approved failed tests.
Fixed.

**Review by paulbalandan — APPROVED** [review]

**Pull request merged**